### PR TITLE
Entry-wise batchnorm

### PR DESCRIPTION
**Pull request by timmoon10**
Our current batchnorm does not work well with large fully-connected networks (like the last few layers in AlexNet) since it only supports data-parallel data. Supporting data in (`MC`, `MR`) format is not feasible. This PR implements an entry-wise batchnorm layer, which is equivalent to our current usage with flattened data, but can also support the model-parallel case. Gradient checking passes on for both data-parallel and model-parallel cases on both CPU and GPU.

Note that we currently only support global statistics. I don't think it would be hard to support local statistics, but it might not be advisable since statistics counts are already so small. Defining reasonable "statistics groups" would also be trickier for (`MC`, `MR`) data: row communicators would probably consist of non-adjacent processes, so I don't expect much latency benefits from reducing over a subset of the row communicator.

### CODE REVIEW

File: include/lbann/layers/regularizers/entrywise_batch_normalization.hpp [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+#ifndef LBANN_LAYERS_REGULARIZERS_ENTRYWISE_BATCH_NORMALIZATION_HPP_INCLUDED
+#define LBANN_LAYERS_REGULARIZERS_ENTRYWISE_BATCH_NORMALIZATION_HPP_INCLUDED
+
+#include "lbann/layers/layer.hpp"
+#include "lbann/models/model.hpp"
+#include "lbann/utils/memory.hpp"
+
+namespace lbann {
+
+template <data_layout Layout, El::Device Device>
```
Comment (ndryden) [inline include/lbann/layers/regularizers/entrywise_batch_normalization.hpp]: A touch of documentation would be good. 😉

File: src/layers/regularizers/entrywise_batch_normalization.cpp [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+  /// @todo Arbitrary group sizes
+  comm.allreduce(batch_statistics,
+                 batch_statistics.RedundantComm(),
+                 El::mpi::SUM);
+  const size_t statistics_count = input.Width();
+
+  // Compute mini-batch statistics from sums
+  if (statistics_count <= 1) {
+    // local_mean already has correct values
+    El::Fill(local_batch_var, DataType{1});
```
Comment (ndryden) [inline src/layers/regularizers/entrywise_batch_normalization.cpp]: This case occurs when the mini-batch size being aggregated over is 1 or less, right? We may want to throw an error instead, since the variance isn't well-defined then.

File: src/layers/regularizers/entrywise_batch_normalization.cpp [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+  }
+
+}
+
+/** @brief Backprop for inference.
+ *
+ *  Assumes forward prop uses running statistics. In other words,
+ *  statistics are independent of input.
+ */
+void bp_inference_impl(DataType epsilon,
```
Comment (ndryden) [inline src/layers/regularizers/entrywise_batch_normalization.cpp]: "Backprop for inference" is an odd thing to say. I think this is just computing dL/dx? We should say something along those lines.

File: src/proto/factories/layer_factory.cpp [hunk 4]
```diff
@@ -571,6 +572,11 @@ std::unique_ptr<Layer> construct_layer(
       return lbann::make_unique<selu_dropout<Layout, Device>>(comm, keep_prob);
     }
   }
+  if (proto_layer.has_entrywise_batch_normalization()) {
+    const auto& params = proto_layer.entrywise_batch_normalization();
+    /// @todo Support GPU implementation
```
Comment (ndryden) [inline src/proto/factories/layer_factory.cpp]: Doesn't this support GPUs already?

File: include/lbann/layers/regularizers/entrywise_batch_normalization.hpp (hunk 1 above)
Comment (timmoon10) [reply include/lbann/layers/regularizers/entrywise_batch_normalization.hpp]: It's self-documenting, ain't it? :p

File: src/layers/regularizers/entrywise_batch_normalization.cpp (hunk 2 above)
Comment (timmoon10) [reply src/layers/regularizers/entrywise_batch_normalization.cpp]: I think that's too harsh, since the user will typically have little control on the size of the last mini-batch.

File: src/layers/regularizers/entrywise_batch_normalization.cpp (hunk 3 above)
Comment (timmoon10) [reply src/layers/regularizers/entrywise_batch_normalization.cpp]: I'd like to make a distinction between the implementation for training and for inference. Perhaps more documentation is in order.

File: src/proto/factories/layer_factory.cpp (hunk 4 above)
Comment (timmoon10) [reply src/proto/factories/layer_factory.cpp]: Oops, forgot to delete that.
